Src/PerformanceCollector/Shared/Implementation/WebAppPerformanceCollector/WebAppPerformanceCollector.cs

```diff
@@ -0,0 +1,233 @@
+            }
+
+            this.RegisterPerformanceCounter(
+                pcd.OriginalString,
+                pcd.ReportAs,
+                pcd.CategoryName,
+                pcd.CounterName,
+                pcd.InstanceName,
+                pcd.UsesInstanceNamePlaceholder,
+                pcd.IsCustomCounter);
+        }
+
+        /// <summary>
+        /// Collects a value for a single counter.
+        /// </summary>
+        private static double CollectCounter(string coutnerOriginalString, ICounterValue counter)
+        {
+            try
+            {
+                return counter.GetValueAndReset();
+            }
+            catch (Exception e)
+            {
+                 throw new InvalidOperationException(
+                    string.Format(
+                        CultureInfo.CurrentCulture,
+                        Resources.WebAppPerformanceCounterReadFailed,
+                        coutnerOriginalString),
+                    e);
+            }
+        }
+
+        /// <summary>
+        /// Register a performance counter for collection.
+        /// </summary>
+        private void RegisterPerformanceCounter(string originalString, string reportAs, string categoryName, string counterName, string instanceName, bool usesInstanceNamePlaceholder, bool isCustomCounter)
+        {
+            ICounterValue counter = null;
+
+            try
+            {
+                counter = this.factory.GetCounter(originalString, reportAs);
+            }
+            catch
+            {
+                PerformanceCollectorEventSource.Log.CounterNotWebAppSupported(originalString);
+                return;
+            }
+
+            bool firstReadOk = false;
+
+            try
+            {
+                // perform the first read. For many counters the first read will always return 0
+                // since a single sample is not enough to calculate a value
+                var value = counter.GetValueAndReset();
+                firstReadOk = true;
+            } 
+            catch (Exception e)
+            {
+                throw new InvalidOperationException(
+                    string.Format(
+                        CultureInfo.CurrentCulture,
+                        Resources.WebAppPerformanceCounterFirstReadFailed,
+                        counterName),
+                    e);
+            }
+            finally
+            {
+                PerformanceCounterData perfData = new PerformanceCounterData(
+                        originalString,
+                        reportAs,
+                        usesInstanceNamePlaceholder,
+                        isCustomCounter,
+                        !firstReadOk,
+                        categoryName,
+                        counterName,
+                        instanceName);
+
+                this.performanceCounters.Add(new Tuple<PerformanceCounterData, ICounterValue>(perfData, counter));
```

Comment:
This is not the exception associated with counter being null. This is associated when retrieving the  value is failed, in which case, we treat the counter as a bad state counter